Validate generated user-data yaml with jsonschema

afbjorklund opened the pull request:
Currently there are 3 validation errors being reported:

`jsonschema: '/ca_certs/remove_defaults' does not validate with https://raw.githubusercontent.com/canonical/cloud-init/main/cloudinit/config/schemas/schema-cloud-config-v1.json#/allOf/8/$ref/properties/ca_certs/$ref/properties/remove_defaults/type: expected boolean, but got string`

`jsonschema: '/ca_certs/trusted' does not validate with https://raw.githubusercontent.com/canonical/cloud-init/main/cloudinit/config/schemas/schema-cloud-config-v1.json#/allOf/8/$ref/properties/ca_certs/$ref/properties/trusted/type: expected array, but got null`

`jsonschema: '/users/0' does not validate with https://raw.githubusercontent.com/canonical/cloud-init/main/cloudinit/config/schemas/schema-cloud-config-v1.json#/allOf/49/$ref/properties/users/items/oneOf/0/type: expected string, but got object`

And 1 deprecation warning, not yet being shown here.

```json
        "uid": {
          "description": "The user's ID. Default value [system default]",
          "oneOf": [
            {
              "type": "integer"
            },
            {
              "type": "string",
              "changed": true,
              "changed_description": "The use of ``string`` type is deprecated. Use an ``integer`` instead.",
              "changed_version": "22.3"
            }
          ]
        }

```

https://github.com/canonical/cloud-init/blob/main/cloudinit/config/schemas/schema-cloud-config-v1.json

----

Issue:

* #2265

The first two are fixed by

* #2266

But "ssh-authorized-keys" is a problem.

```json
        "ssh_authorized_keys": {
          "description": "List of SSH keys to add to user's authkeys file. Can not be combined with ``ssh_redirect_user``",
          "type": "array",
          "items": {
            "type": "string"
          },
          "minItems": 1
        },
```

https://git.launchpad.net/cloud-init/commit/?id=b27f713a

afbjorklund commented:
As an example there was a warning in place for ca-certs, something that affected lima earlier (326bbaaeaae30ad6459634ecf6e1a4eb052a8084)

```json
    "cc_ca_certs": {
      "type": "object",
      "properties": {
        "ca_certs": {
          "$ref": "#/$defs/ca_certs.properties"
        },
        "ca-certs": {
          "allOf": [
            {
              "$ref": "#/$defs/ca_certs.properties"
            },
            {
              "deprecated": true,
              "deprecated_version": "22.3",
              "deprecated_description": "Use ``ca_certs`` instead."
            }
          ]
        }
      }
    },
```

But no such deprecation for ssh-authorized-keys.

afbjorklund commented:
The cidata tests are supposed to be failing, until the schema is valid (i.e. `ssh_authorized_keys`)

jandubois commented:
> The cidata tests are _supposed_ to be failing, until the schema is valid (i.e. `ssh_authorized_keys`)

I don't think validation should fail for using deprecated expressions. We cannot assume that all supported distros will have updated to a new `cloud-init` immediately, so Lima needs to continue to use the deprecated form until the new variant is supported everywhere.

This is kind of the point of deprecation instead of outright removal.

jandubois commented on pkg/cidata/schema.go:
```diff
@@ -0,0 +1,40 @@
+	"gopkg.in/yaml.v3"
+)
+
+const schemaURL = "https://raw.githubusercontent.com/canonical/cloud-init/main/cloudinit/config/schemas/schema-cloud-config-v1.json"
+
+//go:embed schema-cloud-config-v1.json
+var schemaText string
```
I don't have time for a full review right now, but it bugs me that the URL points to `HEAD` of `main` in a repo, but we vendor a specific version. Shouldn't the URL match the version we embed?

I'm also not a big fan of manually vendoring files from other repos; I would rather create a git submodule for the other repo to create a fixed dependency to a specific commit.

I haven't tested it, but I assume we could make `schema-cloud-config-v1.json` a symlink into the submodule, a Go would still embed it correctly.

I know some people hate using submodules, so not sure if we want to go that route. Any opinions @AkihiroSuda ?

AkihiroSuda replied:
I don't feel that using submodules is the best approach just for a single JSON file.

Probably we can just copy the JSON file into the repo?

jandubois replied:
> Probably we can just copy the JSON file into the repo?

How do we even know which version of `cloud-init` this file comes from? It is just `HEAD` of `main` at some random point in time.

The name `schema-cloud-config-v1.json` is misleading, as it seems to imply it is a fixed version, but it it being updated all the time.

Eventually we will want to update it. What is the procedure to do it?

AkihiroSuda replied:
`go generate` should wget it and validate the sha256sum?

afbjorklund replied:
I can add a specific version of this file, using "main" was just what https://www.schemastore.org/json/ said

afbjorklund replied:
https://raw.githubusercontent.com/canonical/cloud-init/24.1/cloudinit/config/schemas/schema-cloud-config-v1.json

AkihiroSuda commented on pkg/cidata/schema.go:
```diff
@@ -0,0 +1,40 @@
+	"gopkg.in/yaml.v3"
+)
+
+const schemaURL = "https://raw.githubusercontent.com/canonical/cloud-init/main/cloudinit/config/schemas/schema-cloud-config-v1.json"
```
What's the license of this file?

jandubois replied:
According to https://github.com/canonical/cloud-init/blob/main/LICENSE the repo is under your choice of GPL-3 or Apache-2.

afbjorklund replied:
Will add a license statement for the file.

afbjorklund commented:
> I don't think validation should fail for using deprecated expressions.

It is not deprecated in the jsonschema, it is outright missing (and failing)

afbjorklund commented:
I added cloud-init jsonschema "latest" documentation version 24.1.3

Then I added my cloud-init patch as well: https://github.com/canonical/cloud-init/pull/5162

Note: the latest version on github is 24.1.4

Note: the jsonschema on `main` is different

```diff
index ff61dcaa6..a71cec996 100644
--- a/cloudinit/config/schemas/schema-cloud-config-v1.json
+++ b/cloudinit/config/schemas/schema-cloud-config-v1.json
@@ -835,6 +835,13 @@
                 "vault_password_file": {
                   "type": "string"
                 },
+                "verify_commit": {
+                  "type": "boolean",
+                  "default": false
+                },
+                "inventory": {
+                  "type": "string"
+                },
                 "module_name": {
                   "type": "string"
                 },
```

https://github.com/canonical/cloud-init/commit/b1bfa594b6e0679db6ed00ca83a9ddbecd5610b4

afbjorklund commented:
The patch for deprecating `ssh-authorized-keys` has now been accepted upstream, and is on `main`

afbjorklund commented:
Now when cloud-config.yaml is available in the instance directory, we can use `limactl validate` with it too.

* https://github.com/lima-vm/lima/pull/2271

afbjorklund commented:
It is possible to use the go version through cli, so we don't need to included it with lima...

```
go install github.com/santhosh-tekuri/jsonschema/cmd/jv@latest
```

And then it (`jv`) can be invoked with the schema file, similar to `check-jsonschema` above:

```
jv https://raw.githubusercontent.com/canonical/cloud-init/main/cloudinit/config/schemas/schema-cloud-config-v1.json ~/.lima/default/cloud-config.yaml
```
